compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs

```diff
@@ -1000,6 +1001,51 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
         err.emit()
     }
 
+    fn apply_do_not_recommend<'b>(
+        &self,
+        trait_predicate: &'_ mut ty::Binder<'b, ty::TraitPredicate<'tcx>>,
+        obligation: &'_ mut PredicateObligation<'tcx>,
+    ) -> ty::Binder<'b, ty::TraitPredicate<'tcx>>
+    where
+        'tcx: 'b,
+    {
+        let mut temp_trait_predicate = trait_predicate.clone();
+        let mut base_cause = obligation.cause.code().clone();
+        // we first need to check whether the current impl
+        // is marked as `#[do_not_recommend]`
+        if let ObligationCauseCode::ImplDerivedObligation(ref c) = base_cause {
+            if self.tcx.has_attr(c.impl_or_alias_def_id, sym::do_not_recommend) {
+                use std::ops::Deref;
+
+                let code = c.derived.parent_code.deref().clone();
+                obligation.cause.map_code(|_| code);
+                obligation.predicate = c.derived.parent_trait_pred.to_predicate(self.tcx);
+                temp_trait_predicate = c.derived.parent_trait_pred.clone();
+                *trait_predicate = c.derived.parent_trait_pred;
+            }
+        }
+
+        // later we need to walk the parent tree
```

Comment:
Why does this need to be a separate loop? The logic that applies above (we need to peel until we find the obligation preceding the *first* `do_not_recommend` in the stack) should also apply below -- should be able to just have one loop.